KEP-10189: MultiKueue Global Status Condition (#10266)
kubernetes-sigs/kueue: keps/10189-multikueue-global-status-condition/README.md

- [Drawbacks](#drawbacks)
- [Alternatives](#alternatives)
  - [Global Status Summary](#global-status-summary)
  - [Visibility API](#visibility-api)
  - [Consolidated Workload State](#consolidated-workload-state)
<!-- /toc -->

## Summary

The Workload status in the MultiKueue Manager Cluster must reflect the true state of the workload derived from the Worker Clusters,
including a human-readable message explaining the state (e.g., "Waiting for quota in cluster X").

This KEP focuses on a mechanism to provide such a high-level summary in the form of a new Workload Status Condition - the **MultiKueueWorkload** condition.
The condition:
1. Will be populated for workloads created on the Manager Cluster.
2. Will provide a high-level, human-readable message explaining the state of the Manager Workload.
3. Will provide insights into the Workload's progress throughout its **whole lifecycle**.
4. Will aggregate information from all the Remote Workloads dispatched by MultiKueue to Worker Clusters for the subject Manager Workload.

## Motivation

Currently, the Workload status subresource is missing key information aggregating data from its remote counterparts on Worker Clusters.

> The existing conditions do not track the process of dispatching Remote Workloads to Workers, obfuscating the details of the most critical part of MultiKueue wWrkload's execution.

Reviewer: typo: wWrkload

This forces users to search across all Worker Clusters to be able to see the big picture.

Moreover, it does not natively support a contract defining a human-readable execution status.
It instead relies on a list of Conditions and support information provided inside the WorkloadStatus field for the user to piece the actual global state of the underlying job's execution together on their own.

To amend this, we need a way to present the user with a clearly defined, user-readable summary of the Global State, which aggregates information from across all the clusters of the MultiKueue environment.

If nothing is implemented, the users will remain forced to rely on:
* the conditions of the Manager Workload, which are misaligned and non-representative of all the extensive logic happening in MultiKueue,

> * manually querying and aggregating the conditions of Remote Workload across all Registered Workers; this covers a lot of distributed data; users are missing the core information of which Workers are eligible for distribution by virtue of being put forward by the Distribution Strategy.

Reviewer: So if the second part is clearer than the first one (which I'd agree with), then maybe we don't need the first one?

How about e.g. this (which is just reshuffling of some of your sentences):

_Currently, the Workload status subresource is missing key information aggregating data from its remote counterparts on Worker Clusters. The users are forced to rely on:_

* _the conditions of the Manager Workload, which are misaligned and non-representative of the extensive logic in MultiKueue,_
* _manually querying and aggregating the conditions of Remote Workloads across all registered workers; this covers a lot of distributed data, and users lack core information about which workers are eligible for dispatch by virtue of being put forward by the dispatch strategy._

_To address this, we need a way to present users with a clearly defined, human-readable summary of the global state, which aggregates information from all clusters in the MultiKueue environment._

* **RUNNING** - local workload is admitted (has the admitted condition); a worker was nominated; the remote is admitted (has the admitted condition); the underlying job will attempt to execute; if it finishes we transition into the SUCCESS/FAILED state, otherwise we back-off,
* **WORKER_SELECTED** - local workload is admitted (has the admitted condition); a single worker was nominated; the remote has currently received quota but is not admitted yet;
this is possible if the MultiKueueWaitForWorkloadAdmitted feature is disabled; it can also occur as a consequence of graduating from the WAITING_FOR_WORKER state when the behavior described in  #9338  occurs,
* **WAITING_FOR_WORKER** -  quota reserved on the local workload; dispatching remotes to eligible workers and waiting  to nominate one of them; a worker will be nominated once the remote achieves a state allowing it to graduate to either the WORKER_SELECTED or the RUNNING state,
Note: Local workload can be admitted (have the admitted condition), as described in #9338 (when MultiKueue admits a workload on a worker but then it gets preempted there with the MultiKueueRedoAdmissionOnEvictionInWorker gate enabled).

> * **WAITING_FOR_WORKER_NOMINATION** - specific to a non-primary component workload in the multi-workload-resource handling scenario;  quota reserved on the component workload; the component workload is waiting for the primary to nominate a worker to create a remote on,

Author: For now opted to use the "multi-workload resource type name" Workload (e.g. LeaderWorkerSet Workload) naming scheme - please let me know your thoughts.

* **WAITING_FOR_QUOTA** - local workload is waiting to be granted a quota reservation on the Manager Cluster.

For each MultiKueueGlobalStatus a message - **MultiKueueGlobalStatusMessage** - will be defined:
* SUCCESS: `Workload has finished successfully on worker cluster: <worker cluster reference>.`
* FAILED: `Workload failed after admission on worker cluster: <worker cluster reference>.`
* INACTIVE: `Workload inactive: <reason>.`

> * RUNNING: `Workload admitted on worker cluster: <worker cluster reference>.`

Reviewer 2: We wouldn't lose the `AdmittedInWorker` condition, it would simply remain `True` with the proposed "Workload admitted on worker cluster: <worker cluster reference>." message.

And for "success" - how does it differ from `WorkloadFinished`? It should be kept in sync between the local workload and the running remote so there's not really a distinction of "whether it was finished locally or on a remote worker", no?

> * RUNNING: `Workload admitted on worker cluster: <worker cluster reference>.`

Author: Yes, but then we give the AdmittedInWorker=False a different meaning then when AdmittedInWorker is not present (i.e. mk workload vs non-mk workload). I would prefer keeping "missing" to mean "false".

Still the main issue with using AdmittedInWorker is immediate readability:
|MultiKueueWorkload|AdmittedInWorker|
|---|---|
|Immediately lets the user know the workload is an MK one|Requires user to know the context behind the condition|
|Allows identifying the state of the workload immediately|Requires checking other conditions to determine if wl has finished and with what result|
|Has only two legal states: present (true) or missing (false)|Has 3 states: present (true), present (false) and missing (not even an MK workload)|

> * RUNNING: `Workload admitted on worker cluster: <worker cluster reference>.`

Author: Mentioned in an alternative - ptal.

* WAITING_FOR_QUOTA: `Workload awaiting quota on the manager cluster.`

The **MultiKueueWorkload** condition will be defined as:
|Field|Value|
|---|---|
|Type|_MultiKueueWorkload_|
|Status|_True_|
|Reason|MultiKueueGlobalStatus|
|Message|MultiKueueGlobalStatusMessage|

The condition will be gated behind the **MultiKueueGlobalStatusCondition** feature gate.

### User Stories

#### Story 1

I want to get an idea of when my job will be admitted. Has it already received quota on the manager?
Is it being dispatched? If yes - what are the workers has it been dispatched to, so that I can check the status of some or all of them for how my job is doing there?
Or maybe it is already running on a specific worker? If yes - which one?

#### Story 2

My job hasn’t been admitted for a long time. I would like to verify why that is.
Are remotes being created? Have they been dispatched to all workers? What is their status there? Are they waiting for quota? Are they waiting for admission checks?

### Notes/Constraints/Caveats

This proposal is limited to only providing the user with a high-level summary.
This is a minimum value proposition, lacking in a few key areas:

> * In the WAITING_FOR_WORKER state we only provide a miniscule summary of what is happening on the Workers. In reality, this state is much more complex than the messages we propose can describe and could greatly benefiot form being accompanied by a set of aggregations describing in greater detail what is the status of each Remote Workload.

Reviewer: ```suggestion
* In the WAITING_FOR_WORKER state we only provide a miniscule summary of what is happening on the Workers. In reality, this state is much more complex than the messages we propose can describe and could greatly benefit form being accompanied by a set of aggregations describing in greater detail what is the status of each Remote Workload.
```

  // The underlying job is being executed on said Worker Cluster.
  Running = "RUNNING"

  // WorkerSelected state means the workload has the "Admitted" condition on the Manager Cluster but was not admitted on the Worker yet.
  // A specific Worker was selected, but the workload has only managed to reserve quota there so far.
  WorkerSelected = "WORKER_SELECTED"

  // WaitingForWorker state means the workload has received quota on the Manager Cluster.
  // MultiKueue is currently dispatching remote workloads to eligible Workers.
  WaitingForWorker = "WAITING_FOR_WORKER"

  // WaitingForWorkerNomination state is specific to a non-primary component workload in the multi-workload-resource handling scenario.
  // It means the component workload has received quota reservation on the Manager
  // and is waiting for the primary component workload to select a worker to dispatch a remote to.
  WaitingForWorkerNomination = "WAITING_FOR_WORKER_NOMINATION"

  // WaitingForQuota state means the workload is currently in the "Pending" state on the Manager Cluster (does not currently hold any quota reservations).

>   WaitingForQuota = "WAITING_FOR_MANAGER_QUOTA"

Reviewer: ```suggestion
  // WaitingForManagerQuota state means the workload is currently in the "Pending" state on the Manager Cluster (does not currently hold any quota reservations).
  WaitingForManagerQuota = "WAITING_FOR_MANAGER_QUOTA"
```

* _worker selected by the primary local workload_ and _primary local workload reference_ are already identified in the current reconciliation process,
* _multi-workload resource type name_ can be determined by verifying the type of the underlying job.

### Test Plan

[ ] I/we understand the owners of the involved components may require updates to
existing tests to make this code solid enough prior to committing the changes necessary
to implement this enhancement.

#### Unit tests

Unit tests will be added for the methods identifying the MultiKueueGlobalStatus and calculating the condition as part of the [MultiKueue Workload Controller](https://github.com/Singularity23x0/kueue/blob/66c2cf1a93ee3f19d8b29309d27832a4595d4106/pkg/controller/admissionchecks/multikueue/workload_test.go#L17) test suite.

#### Integration tests

Integration tests for assigning the expected condition correctly for each Global Status will be added in a new file as part of the [MultiKueue Integration Test Suite](https://github.com/Singularity23x0/kueue/blob/66c2cf1a93ee3f19d8b29309d27832a4595d4106/test/integration/multikueue/suite_test.go#L93).

#### e2e tests

> Integration tests for assigning the expected condition correctly for each Global Status will be added in a new file as part of the [MultiKueue E2E Test Suite](https://github.com/Singularity23x0/kueue/blob/66c2cf1a93ee3f19d8b29309d27832a4595d4106/test/e2e/multikueue/suite_test.go#L62).

Reviewer 3: `Integration` -> `End to end`

| WAITING FOR WORKER | Local has the admission but not the admitted condition. Dispatching remotes to one or more workers. All remotes are PENDING (Pending) or ADMISSION PENDING (QuotaReserved). | — | WAITING_FOR_WORKER | — |
| WAITING FOR WORKER NOMINATION | Local workload is a non-primary workload in a group of composite workloads. Local has got an admission but not the admitted condition. Primary has not selected a worker yet. | — | WAITING_FOR_WORKER_NOMINATION | — |
| PENDING | Local workload does not have an admission. No remotes exist. | Workload does not have an admission. | WAITING_FOR_MANAGER_QUOTA | Pending |

This consolidated set of statuses could then be used to populate a new, stand-alone field in the WorkloadStatus, allowing for a generic solution providing a high-level status summary for both individual and MultiKueue workloads.

The main caveat here is that this approach risks being confusing to users, as the Workload States shift meanings significantly depending on whether the Workload is a Manager Workload or an Individual (remote or non-MultiKueue) Workload.

### Granular Conditions

> Instead of defining a single Condition, we could provide one or more Granular Conditions, each servicing the purpose of supplying data missing from the WorkloadStatus, allowing to piece the Global Status of a MultiKueue workload by the user on their own.

Reviewer: ```suggestion
Instead of defining a single Condition, we could provide one or more Granular Conditions, each serving the purpose of supplying data missing from the WorkloadStatus, allowing to piece the Global Status of a MultiKueue workload by the user on their own.
```

Outside of that, it closely resembles the proposed solution, especially given that both Reason and Message values can be copied directly in all cases but the INACTIVE state one.

#### Condition per status

We can define multiple conditions providing users with data they lack to establish the Global Status of a MultiKueue workload.

These conditions could be one or more of:
* AdmittedInWorker - set to true when in the RUNNING status. The message would identify the worker.
* QuotaReservedInWorker - set to true when in the WORKER_SELECTED status. The message would identify the worker.
* WorkerNominated - set to False when no worker was nominated for dispatch yet. Set to True once a single worker is nominated. This can mean entering the RUNNING or WORKER_SELECTED states, as well as cover exiting the WAITING_FOR_WORKER_NOMINATION state in case of non-primary composite workloads. The reason and message would identify which of these 3 cases we encountered.
* MultiKueueManagerWorkload - present and set to True only for MultiKueue Manager Workloads, ensuring we are always aware if the workload in question is a Manager Workload or not.

#### Disadvantages

> The main issue here, is that while we ensure the status covers all MultiKueue cases, reading the state of the job's execution remains complex. We still have to read an understand a multitude of conditions and status information to piece together the inner workings of MultiKueue, isntead of being provided with a clear, human-readable summary. This is especially apparrent when considering the latter variant, where the number of added conditions expands to 3 or 4.

Reviewer: ```suggestion
The main issue here is that, while we ensure the status covers all MultiKueue cases, reading the state of the job's execution remains complex. We still have to read and understand a multitude of conditions and status information to piece together the inner workings of MultiKueue, isntead of being provided with a clear, human-readable summary. This is especially apparent when considering the latter variant, where the number of added conditions expands to 3 or 4.
```

> The main issue here, is that while we ensure the status covers all multi-kueue cases, reading the state of the job's execution remains complex. We still have to read an understand a multitude of conditions and status information to piece together the inner workings of MultiKueue, isntead of being provided with a clear, human-readable summary. This is especially apparrent when considering the latter variant, where the number of added conditions expands to 3 or 4.

Reviewer: ```suggestion
The main issue here, is that while we ensure the status covers all MultiKueue cases, reading the state of the job's execution remains complex. We still have to read an understand a multitude of conditions and status information to piece together the inner workings of MultiKueue, isntead of being provided with a clear, human-readable summary. This is especially apparrent when considering the latter variant, where the number of added conditions expands to 3 or 4.
```

  * Non-primary component workload:
  `<multi-workload resource type name> Workload awaiting admission on Worker Cluster: <worker selected by the primary local workload>, selected by the <multi-workload resource type name> primary workload: <primary local workload reference>.`
* WAITING_FOR_WORKER_NOMINATION: `<multi-workload resource type name> Workload waiting for <primary local workload reference> to select a Worker Cluster.`
* WAITING_FOR_MANAGER_QUOTA: `Workload awaiting quota on the Manager Cluster.`

The **MultiKueueWorkload** condition will be defined as:
|Field|Value|
|---|---|
|Type|_MultiKueueWorkload_|
|Status|_True_|
|Reason|MultiKueueGlobalStatus|
|Message|MultiKueueGlobalStatusMessage|

The condition will be gated behind the **MultiKueueGlobalStatusCondition** feature gate.

### User Stories

#### Story 1

I want to get an idea of when my job will be admitted. Has it already received quota on the manager?
Is it being dispatched? If yes - what are the workers it has been dispatched to, so that I can check the status of some or all of them for how my job is doing there?

> Or maybe it is already running on a specific worker? If yes - which one?

Reviewer 2: I think all those questions can be answered already, even without the proposed condition. So maybe this story could put more emphasis on the centralization of this information, if that's the goal.

Author: Expanded - ptal.

The exact MultiKueueGlobalStatus to assign can be determined using the conditions in the Manager Workload and the conditions in each of the Remote Workloads, all already gathered as part of the reconciliation process.

For the defined messages, each variable can be determined using the data present during the reconciliation process:
* _worker cluster reference_ of the selected Worker is retrieved as part of processing (when in one of {SUCCESS, FAILED, RUNNING, WORKER_SELECTED} states),
* the _reason_ for why the Manager Workload is inactive is present in an appropriate condition of the Manager Workload,
* _number of remotes_ is number of the noted remotes,
* _worker selected by the primary local workload_ and _primary local workload reference_ are already identified in the current reconciliation process,
* _multi-workload resource type name_ can be determined by verifying the type of the underlying job.

### Test Plan

[ ] I/we understand the owners of the involved components may require updates to
existing tests to make this code solid enough prior to committing the changes necessary
to implement this enhancement.

#### Unit tests

> Unit tests will be added for the methods identifying the MultiKueueGlobalStatus and calculating the condition as part of the [MultiKueue Workload Controller](https://github.com/Singularity23x0/kueue/blob/66c2cf1a93ee3f19d8b29309d27832a4595d4106/pkg/controller/admissionchecks/multikueue/workload_test.go#L17) test suite.

Reviewer 2: You linked to your fork repo. I'd change it to the kubernetes-sigs one.

To address this, we need a way to present users with a clearly defined, human-readable summary of the global state, which aggregates information from all clusters in the MultiKueue environment.

### Goals

* Adding a new Workload Status Condition describing the current, actual state of the Manager Workload.
* Defining an enumeration of mutually exclusive MultiKueue Manager Workload states covering the whole lifecycle of the workload.
* Defining a set of high-level, human-readable messages to accompany the enumerated states, describing them further.
* Populating the condition for relevant workloads.

### Non-Goals

* Defining Status Summary describing data structures beyond what's necessary for implementing the new Workload Status Condition.

## Proposal

The lifecycle of the Manager Workload will be split into the following **MultiKueueGlobalStatus** enumeration:
* **SUCCESS** - workload finished successfully,

> * **FAILED** - workload finished with failed state,

Reviewer 4: 1. Why do we need these reasons even though the Workload already has the duplicated Finished Workload condition?
2. How to keep consistency between the Finished condition type and the new ones?

Author: 1. I initially addressed this matter in the "Granular Conditions" alternatives section. TLDR: the idea is to provide this information via a unified, human-readable message.
2. The new condition will be populated in the MultiKueue Core Workload Reconciler - there we can ensure that if the finished condition is present and true, our new condition is compliant.

Author: To expand on 1: While the data is available, it requires a lot of analysis to gather into a simple, readable message. The core idea behind this Condition is to provide users with such a simple, readable message to quickly summarize the state of the workload.

Author: For added generic context, please refer to the Motivation section.

Reviewer 4: The Workload condition is responsible for the Workload condition (phase) atomically.
I'm worried about adding a new condition type only for MultiKueue because Workloads could be in an unstable phase, which means the Workload terminal state will be split into 2 condition types.

For example, as a first step, the Workload gets the `type: MultiKueueWorkload, reason: SUCCESS`, but the Workload doesn't have `type: Finished` condition even though the Workload is actually in the terminal phase.

Also, this proposed pattern loses the status field (`conditions[*].status: True|False`) responsibility.
For example, what does `type: MultiKueueWorkload, reason: SUCCESS, status: False` mean? I imagine that this state doesn't have any meaningful information.

So, I'm wondering if (1. we can have a MultiKueue specific status instead of conditions or (2. Expand existing condition reasons like `{type: Finished: reason: SucceededOnMultiKueueWorkerCluster}`.

I guess that 1 could align with your use cases (emphasize MultiKueue phase <status>). But, (1. has a risk where we have to imagine the future API design and decide the best API at this time.

The safest approach is introducing (1 as a first step, then in the second step, we can move (1) to a dedicated API (2) based on feedback. But I want to decide if we should select this step-by-step approach based on MultiKueue roadmaps. So, could you share the MK roadmap if you can?

Author: > For example, as a first step, the Workload gets the type: MultiKueueWorkload, reason: SUCCESS, but the Workload doesn't have type: Finished condition even though the Workload is actually in the terminal phase. 

The intent is to update the MultiKueueWorkload condition on every MK Reconcile, which would be triggered on a change to the Finished condition. This would ensure the MK Condition is always synced.

> Also, this proposed pattern loses the status field (conditions[*].status: True|False) responsibility. For example, what does type: MultiKueueWorkload, reason: SUCCESS, status: False mean? I imagine that this state doesn't have any meaningful information.

Indeed, that status would not be utilized. This is one of the downsides of this solution as opposed to encoding the aggregated MK status somewhere else.

> So, I'm wondering if (1. we can have a MultiKueue specific status instead of conditions or (2. Expand existing condition reasons like {type: Finished: reason: SucceededOnMultiKueueWorkerCluster}.

Re. 1. Do you mean to add a new field to the Wl Status field? Or placing it in a new resource (similar to the Alt. 1 proposal)?

> The safest approach is introducing (1 as a first step, then in the second step, we can move (1) to a dedicated API (2) based on feedback. But I want to decide if we should select this step-by-step approach based on MultiKueue roadmaps. So, could you share the MK roadmap if you can?

The plan was to:
A. Introduce the MK Status Summary (current proposal was to do it as a Condition to make the implementation simpler, without adding the need to modify the resources and deciding on the API already).
B. At a later stage - reevaluate the need for introducing a more robust summary for MK as mentioned in the Alternatives section, potentially moving the MultiKueueGlobalStatus form the conditions into there alongside a more expansive aggregation of data from across the workers.

Reviewer 4: > > So, I'm wondering if (1. we can have a MultiKueue specific status instead of conditions or (2. Expand existing condition reasons like {type: Finished: reason: SucceededOnMultiKueueWorkerCluster}.
> 
> Re. 1. Do you mean to add a new field to the Wl Status field? Or placing it in a new resource (similar to the Alt. 1 proposal)?

I meant that (1 indicates the Workload status field has a new MultiKueue specific state field.

> > The safest approach is introducing (1 as a first step, then in the second step, we can move (1) to a dedicated API (2) based on feedback. But I want to decide if we should select this step-by-step approach based on MultiKueue roadmaps. So, could you share the MK roadmap if you can?
> 
> The plan was to:
> A. Introduce the MK Status Summary (current proposal was to do it as a Condition to make the implementation simpler, without adding the need to modify the resources and deciding on the API already).
> B. At a later stage - reevaluate the need for introducing a more robust summary for MK as mentioned in the Alternatives section, potentially moving the MultiKueueGlobalStatus form the conditions into there alongside a more expansive aggregation of data from across the workers.

Thank you for sharing that. In that case, if you want to accelerate this proposal for 0.18 release, I'm ok with condition approach.
But, before beta graduation, we have to add the MultiKueue-specific status field in the Workload Custom Resource which means new MultiKueue specific Workload conditions added at this time have to be removed in second alpha stage.

In other words, if we don't move MultiKueue specific Workload state, we never graduate this KEP to Beta.

The following is the proposed timeline:

Timeline (A):

1st alpha (0.18): Add algorithms and condition types and reasons.
2nd alpha (0.19 or 0.20): Add MultiKueue-specific workload state field.
beta (0.20 or 0.21): Enabling state operation by default.
ga ....

Timeline (B):

alpha (0.18 or 0.19): Add MultiKueue-specific workload state field (we never introduce conditions).
beta (0.19 or 0.20): Enabling state operation by default.
ga ....

@Singularity23x0 Which timeline do you prefer?

Author: Proceeding with Timeline (A) - amending the KEP.

Author: Updated - PTAL.

* **INACTIVE** - Manager Workload marked as inactive, preventing it from being scheduled; we can enter this state from any other as the workload can be deactivated both by Kueue and by the user,

> * **RUNNING** - local workload is admitted (has the admitted condition); a single worker was selected; the remote is admitted (has the admitted condition); the underlying job will attempt to execute; if it finishes we transition into the SUCCESS/FAILED state, otherwise we back-off,

Reviewer 4: This intermediate condition is very dangerous because it could cause the statue handling bug.
I strongly wouldn't recommend this reason. Especially, it's very challenging to define what " Running " is.

Additionally, what is the difference between "Admitted" and "Running" condition on the manager cluster?
AFAIK, the Admitted condition on the worker-cluster will be populated to the manager cluster, then the manager cluster will add "Admitted" condition to the Workload on the manager cluster after AdmissionCheck=Ready state.

I understand that this doesn't indicate the underlying Pods are running in the nominated worker cluster.
But, in general, the "Running" is estimated to indicate that actual application processes are executed.

I see, this indicates the consolidated condition reason based on Admitted + `clusterName != ""`.
So, instead of "Running", I would propose "AdmittedOnWorkerCluster" as an alternative.

Author: The way the "Admitted" Condition works on the Manager is not based on the Worker Workload directly.
The manager workload sets Admitted=True when it gets quota reservation on the manager and all the ACs are Ready. The MultiKueue AC will be Ready either when we get Admitted or get QuotaReserved on one of the Workers, depending on whether MultiKueueWaitForWorkloadAdmitted is enabled.

I wanted to avoid using "ADMITTED" so as not to confuse it with the Admitted=True condition.

The split between WORKER_SELECTED and RUNNING stems form the MultiKueueWaitForWorkloadAdmitted feature, where we can have Admitted=True and `clusterName != ""`, but:
* MultiKueueWaitForWorkloadAdmitted DISABLED - the Remote is not yet admitted, so we say WORKER_SELECTED,
* MultiKueueWaitForWorkloadAdmitted ENABLED or we were in WORKER_SELECTED  and now the Remote Workload is actually Admitted on the Worker.

Renaming to "ADMITTED_ON_WORKER_CLUSTER" or "ADMITTED_ON_WORKER" sounds like an ok alternative.

Author: WORKER_SELECTED is no longer a concern.
ADMITTED_ON_WORKER added.

> * **WORKER_SELECTED** - local workload is admitted (has the admitted condition); a single worker was selected; the remote has currently received quota but is not admitted yet;

Reviewer 4: Why do we need this? This is obviously duplicated with `.status.clusterName`.

Author: To differentiate between being Pending, with Quota Reserved or Admitted on the worker cluster.
When clusterName is set:
* WAITING_FOR_WORKER means we are still pending on the worker,
* WORKER_SELECTED means we have quota but are not yet Admitted there,
* RUNNING means we are admitted on the Worker.

Author: Aside from that, the core idea of providing a unified, human readable summary of the status applies here.
See: https://github.com/kubernetes-sigs/kueue/pull/10266#discussion_r3117908206

this is possible if the MultiKueueWaitForWorkloadAdmitted feature is disabled,

> * **WAITING_FOR_WORKER** -  quota reserved on the local workload; dispatching remotes to nominated workers and waiting for one of them to be selected - a worker will be selected once the remote achieves a state allowing it to graduate to either the WORKER_SELECTED (receiving quota reservation, MultiKueueWaitForWorkloadAdmitted feature gate disabled) or the RUNNING (workload receives the admitted condition set to true) state,

Reviewer 4: Why do we need this? This is obviously deplicated with `len(.status.nominatedClusterNames) != 0 && .status.clusterName == ""`.

Author: See: https://github.com/kubernetes-sigs/kueue/pull/10266#discussion_r3117908206

> * **WAITING_FOR_WORKER_NOMINATION** - specific to a non-primary component workload in the multi-workload-resource handling scenario;  quota reserved on the component workload; the component workload is waiting for the primary to select a worker to create a remote on,

Reviewer 4: Why do we need this? Isn't this duplicated with `len(.status.nominatedClusterNames) == 0 `?
Also, what is "component workload" ?

Author: Why do we need this: see https://github.com/kubernetes-sigs/kueue/pull/10266#discussion_r3117908206

Component workloads denotes all workloads created by multiworkload resources (for example LeaderWorkerSet). They are handled by a MultiWorkloadAdapted in MultiKueue. TLDR:
* multiple workloads are created for such a multiworklaod resource,
* one of the workloads is denoted as Primary,
* the primary works like any other mk workload,
* other component workloads will not have any workers nominated; instead, once the primary finds a suitable Worker, that worker will be nominated for all other component workloads.

> * **WAITING_FOR_MANAGER_QUOTA** - local workload is waiting to be granted a quota reservation on the Manager Cluster.

Reviewer 4: Why do we need this? It's duplicated with "QuotaReserved" condition type.
What is the difference?

Author: Core reason: https://github.com/kubernetes-sigs/kueue/pull/10266#discussion_r3117908206
This status unifies the cases of "quota reserved" being either missing or set to false.

* In an effort to keep the condition message concise, we are limited in how much data relevant to the user we can provide.

In the [alternatives section](#global-status-summary) we briefly describe an expanded Global Status Summary proposal to be revisited in the future if the new condition by itself is deemed insufficient.

### Risks and Mitigations

The main risk of the feature is its effect on the MultiKueue Reconciler's performance and code complexity.
Implementing the condition calculating logic efficiently will require careful consideration of where the Reconciler can reliably determine the state of the Manager Workload.
That may necessitate distributing pieces of logic across functions, complicating the already convoluted logic. Mitigating that will require careful naming and code grouping in the implementation phase.

Another issue stems from the MultiKueue's complex logic itself. It may prove difficult to reliably track the actual state due to the branching nature of the functions.
Countering that will require thorough test coverage, with a special emphasis on implementing a wide coverage in E2E tests.

> ## Design Details

Reviewer 4: @Singularity23x0 I would request to add the state transition diagram either by image or a marmaid graph.
Because we should evaluate Reasons if there are any buggy paths or duplicated and unnecessary reasons.

Author: Added - ptal.

> The new condition type will be defined alongside existing ones in the [workload types file](https://github.com/kubernetes-sigs/kueue/blob/9e20f44bb05758976744a5ee6aa77309677bb22f/apis/kueue/v1beta1/workload_types.go#L628).

Reviewer 4: Could you replace this link with v1beta2?
If you are aiming to add these APIs only to v1beta1, please let me know. In that case, we should discuss API migration and graduation.

```go
const (
  // ...

  // MultiKueueWorkload means the workload is a MultiKueue Workload created on a Manager Cluster.
  // The possible reasons depend on the state of the MK Workload:
  // - SUCCESS,
  // - FAILED,
  // - INACTIVE,
  // - RUNNING,
  // - WORKER_SELECTED,
  // - WAITING_FOR_WORKER,
  // - WAITING_FOR_WORKER_NOMINATION,
  // - WAITING_FOR_MANAGER_QUOTA.
  MultiKueueWorkload = "MultiKueueWorkload"

  // ...
)
```

> The MultiKueueGlobalStatus enumeration will be defined in the [MultiKueue types file](https://github.com/kubernetes-sigs/kueue/blob/9e20f44bb05758976744a5ee6aa77309677bb22f/apis/kueue/v1beta1/multikueue_types.go#L17).

Reviewer 4: ditto about link

> ```go

Reviewer 4: When we define reason, we should use CamelCase. For example, the "WORKER_SELECTED" should be "WorkerSelected".

  // A specific Worker was selected, but the workload has only managed to reserve quota there so far.
  WorkerSelected = "WORKER_SELECTED"

  // WaitingForWorker state means the workload has received quota on the Manager Cluster.
  // MultiKueue is currently dispatching remote workloads to eligible Workers.
  WaitingForWorker = "WAITING_FOR_WORKER"

  // WaitingForWorkerNomination state is specific to a non-primary component workload in the multi-workload-resource handling scenario.
  // It means the component workload has received quota reservation on the Manager
  // and is waiting for the primary component workload to select a worker to dispatch a remote to.
  WaitingForWorkerNomination = "WAITING_FOR_WORKER_NOMINATION"

  // WaitingForManagerQuota state means the workload is currently in the "Pending" state on the Manager Cluster (does not currently hold any quota reservations).
  WaitingForManagerQuota = "WAITING_FOR_MANAGER_QUOTA"
)

```

The condition will be populated inside the [Reconciler of the MultiKueue Core Controller](https://github.com/kubernetes-sigs/kueue/blob/9e20f44bb05758976744a5ee6aa77309677bb22f/pkg/controller/admissionchecks/multikueue/workload.go#L158).

> This reconciler already gathers all the necessary data in the form of the [Workload Group](https://github.com/kubernetes-sigs/kueue/blob/9e20f44bb05758976744a5ee6aa77309677bb22f/pkg/controller/admissionchecks/multikueue/workload.go#L83) internal structure.

Reviewer 4: ```suggestion
The condition will be populated inside the Reconciler of the MultiKueue Core Controller.
This reconciler already gathers all the necessary data in the form of the Workload Group internal structure.
```

Could you avoid attaching the code base? Because the implementations will be drastically changed in the future, but this documentation will be kept even in the future.

The exact MultiKueueGlobalStatus to assign can be determined using the conditions in the Manager Workload and the conditions in each of the Remote Workloads, all already gathered as part of the reconciliation process.

For the defined messages, each variable can be determined using the data present during the reconciliation process:
* _worker cluster reference_ of the selected Worker is retrieved as part of processing (when in one of {SUCCESS, FAILED, RUNNING, WORKER_SELECTED} states),
* the _reason_ for why the Manager Workload is inactive is present in an appropriate condition of the Manager Workload,
* _number of remotes_ is number of the noted remotes,
* _worker selected by the primary local workload_ and _primary local workload reference_ are already identified in the current reconciliation process,
* _multi-workload resource type name_ can be determined by verifying the type of the underlying job.

### Test Plan

[ ] I/we understand the owners of the involved components may require updates to
existing tests to make this code solid enough prior to committing the changes necessary
to implement this enhancement.

#### Unit tests

> Unit tests will be added for the methods identifying the MultiKueueGlobalStatus and calculating the condition as part of the [MultiKueue Workload Controller](https://github.com/kubernetes-sigs/kueue/blob/9e20f44bb05758976744a5ee6aa77309677bb22f/pkg/controller/admissionchecks/multikueue/workload_test.go#L17) test suite.

Reviewer 4: ```suggestion
Unit tests will be added for the methods identifying the MultiKueueGlobalStatus and calculating the condition as part of the MultiKueue Workload Controller test suite.
```

ditto about code attachment

#### Integration tests

> Integration tests for assigning the expected condition correctly for each Global Status will be added in a new file as part of the [MultiKueue Integration Test Suite](https://github.com/kubernetes-sigs/kueue/blob/9e20f44bb05758976744a5ee6aa77309677bb22f/test/integration/multikueue/suite_test.go#L93).

Reviewer 4: ```suggestion
Integration tests for assigning the expected condition correctly for each Global Status will be added in a new file as part of the MultiKueue Integration Test Suite.
```

ditto about code attachment

#### e2e tests

> E2E Tests for assigning the expected condition correctly for each Global Status will be added in a new file as part of the [MultiKueue E2E Test Suite](https://github.com/kubernetes-sigs/kueue/blob/9e20f44bb05758976744a5ee6aa77309677bb22f/test/e2e/multikueue/suite_test.go#L62).

Reviewer 4: ```suggestion
E2E Tests for assigning the expected condition correctly for each Global Status will be added in a new file as part of the MultiKueue E2E Test Suite.
```

ditto about code attachment.

This consolidated set of statuses could then be used to populate a new, stand-alone field in the WorkloadStatus, allowing for a generic solution providing a high-level status summary for both individual and MultiKueue workloads.

The main caveat here is that this approach risks being confusing to users, as the Workload States shift meanings significantly depending on whether the Workload is a Manager Workload or an Individual (remote or non-MultiKueue) Workload.

### Granular Conditions

Instead of defining a single Condition, we could provide one or more Granular Conditions, each serving the purpose of supplying data missing from the WorkloadStatus, allowing to piece the Global Status of a MultiKueue workload by the user on their own.

#### AdmittedInWorker Condition

|Corresponding MultiKueueGlobalStatus|AdmittedInWorker Condition Value|
|---|---|
|SUCCESS|nil|
|FAILED|nil|
|INACTIVE|**State**: False, **Reason**: WorkloadInactive|
|RUNNING|**State**: True|
|WORKER_SELECTED|**State**: False|
|WAITING_FOR_WORKER|**State**: False|
|WAITING_FOR_WORKER_NOMINATION|**State**: False|

> |WAITING_FOR_MANAGER_QUOTA|**State**: False|

Reviewer: There's one thing in this table that makes me confused: 

Would you set "Reason" _only_ in the INACTIVE case?

My understanding from our discussion was that (while we may choose to provide the Reason in Inactive case, just for clarity) the other "False" cases (Worker Selected, Waiting for Worker etc.) would need the Reason _even more_, as these are the cases that can't be distinguished based on currently existing manager-side Conditions. Or am I getting it wrong?

Also, the text below "Reason and Message values can be copied directly in all cases but the INACTIVE state one" doesn't help me. The table currently seems to copy Reason for INACTIVE, not for others :)

Author: The idea is that if reason an message are not provided in the table, they are the same as in the core proposal.
